src/kdalgorithms.h

```diff
@@ -585,10 +585,69 @@ auto partitioned(Container &&container, UnaryPredicate predicate)
         std::forward<Container>(container), std::forward<UnaryPredicate>(predicate));
 }
 
+// -------------------- multi_partitioned --------------------
+#if __cplusplus >= 202002L
+namespace detail {
+    template <typename ResultContainer, typename InputContainer, typename KeyFunction>
+    constexpr bool KeyFunctionMapsContainerValueToMapKey =
+        std::is_convertible_v<detail::invoke_result_t<KeyFunction, ValueType<InputContainer>>,
+                              typename ResultContainer::key_type>;
+
+    template <typename ResultContainer, typename InputContainer>
+    constexpr bool InputValueCanConvertToResultContainerValue =
+        std::is_convertible_v<ValueType<InputContainer>,
+                              typename ResultContainer::mapped_type::value_type>;
+}
+#endif
+
+template <typename ResultContainer, typename InputContainer, typename KeyFunction>
+#if __cplusplus >= 202002L
+requires std::is_invocable_v<KeyFunction, ValueType<InputContainer>> && detail::
+    KeyFunctionMapsContainerValueToMapKey<ResultContainer, InputContainer, KeyFunction> && detail::
+        InputValueCanConvertToResultContainerValue<ResultContainer, InputContainer>
+#endif
+auto multi_partitioned(InputContainer &&container, KeyFunction keyFunction)
+{
+    ResultContainer result;
+    auto iterators = read_iterator_wrapper(std::forward<InputContainer>(container));
+    for (auto it = iterators.begin(); it != iterators.end(); ++it) {
```

Comment:
range-based for loop?

Reply:
I'm using a the iterators returned by the read_iterators_wrapper (which is either a normal const iterator, or a move iterator). I don't see how I can change that into a range based for loop, but by all means please enlighten me.